File: BInary/CountOccurence.cpp
// Lower bound case
int lower(vector<int> &arr, int n, int k)
{
    int low = 0;
    int high = n - 1;
    int firstOccur = -1;
    while (low <= high)
    {
        int mid = (low + high) / 2;
        if (arr[mid] == k)
        {
            high = mid - 1;
            firstOccur = mid;
        }
        else if (arr[mid] > k)
        {
            high = mid - 1;
        }
        else
        {
            low = mid + 1;
        }
    }
    return firstOccur;
}

// Upper bound case
int upper(vector<int> &arr, int n, int k)
{
    int low = 0;
    int high = n - 1;
    int lastOccur = -1;
    while (low <= high)
    {
        int mid = (low + high) / 2;
        if (arr[mid] == k)
        {
            lastOccur = mid;
            low = mid + 1;
        }
        else if (arr[mid] > k)
        {
            high = mid - 1;
        }
        else
        {
            low = mid + 1;
        }
    }
    return lastOccur;
}

int count(vector<int> &arr, int n, int k)
{
    // Write Your Code Here
    int firstOccur = lower(arr, n, k);

    // If number is not found then we don't need to find last occurence because it not exist
    if (firstOccur == -1)
        return 0;

    int lastOccur = upper(arr, n, k);

    // subtract first occurence from last ocuurence and add one to it.
    return lastOccur - firstOccur + 1;
}

File: Binary_Search/CountOccurence.java
class CountOccurence {
    int lowerBound(int[] nums, int target) {
        int low = 0;
        int high = nums.length - 1;
        int first = -1;
        while(low <= high) {
            int mid = (low + high) / 2;
            if(nums[mid] == target) {
                first = mid;
                high = mid - 1;
            } else if(nums[mid] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return first;
    }

    int upperBound(int[] nums, int target) {
        int low = 0;
        int high = nums.length - 1;
        while(low <= high) {
            int mid = (low + high) / 2;
            if(nums[mid] == target) {
                low = mid + 1;
            } else if(nums[mid] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return low;
    }

    int countFreq(int[] nums, int target) {
        // code here
        int first = -1;
        int last = -1;

        first = lowerBound(nums, target);

        if(first == -1) {
            return 0;
        } else {
            last = upperBound(nums, target);
        }
        return last - first;
    }
}